### PR TITLE
Add listInfos method to API

### LINKED ISSUE #6097: Get podman machine info from extension
### Is your enhancement related to a problem? Please describe

From an extension, we would like to have memory/cpu information about podman machines in use.

### Describe the solution you'd like

There is a method in the api `containerEngine.info(engineId)`, which provides this information. 

To get the `engineId`, I understand that we nned to execute a `list*` command and, if there are elements returned by any of these functions, to get the engineId from this result (for example, an image returned by `listImages` conatains the engineId on which it is present.

It would be nice to have a more direct way to get the `engineId`, for example whan we have a `ProviderContainerConnection` value (obtained from `provider.getContainerConnections`).

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by feloy**
### What does this PR do?

Add a listInfos to the API, for the extension to get cpu/memory/... information about different container providers.

TODO:
- [x] doc new method
- [x] unit tests

### What issues does this PR fix or reference?

Fixes #6097 

### How to test this PR?

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

**Review by benoitf — COMMENTED** [review]
LGTM

File: packages/main/src/plugin/container-registry.ts [hunk 1]
```diff
@@ -2334,12 +2337,35 @@ export class ContainerProviderRegistry {
...
+        engineType: provider.connection.type,
         cpus: dockerInfo.NCPU,
         memory: dockerInfo.MemTotal,
       };
     }
   }
 
+  async listInfos(options?: containerDesktopAPI.ListInfosOptions): Promise<containerDesktopAPI.ContainerEngineInfo[]> {
+    let providers: InternalContainerProvider[];
+    if (options?.provider === undefined) {
```
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts]:
```suggestion
    if (!options?.provider) {
```

File: packages/main/src/plugin/container-registry.ts [hunk 2]
```diff
@@ -2334,12 +2337,35 @@ export class ContainerProviderRegistry {
...
+      Array.from(providers).map(async provider => {
+        try {
+          return await this.info(provider.id);
+        } catch (error) {
+          console.log('error getting info for engine', provider.name, error);
+          return [];
+        }
+      }),
+    );
+    return infos.flat();
```
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts]: I'm not sure here about the usage of array and then flat
vs filtering out undefined
return Array.from(providers).map(provider => return info(provider).filter(p => !p)

Comment (feloy) [reply packages/main/src/plugin/container-registry.ts]: I cannot get the type checker detect that `filter(info => info)` does not return undefined values. The result is always of type `(containerDesktopAPI.ContainerEngineInfo | undefined)[]`

```
    const infos = await Promise.all(
      Array.from(providers).map(async provider => {
        try {
          return await this.info(provider.id);
        } catch (error) {
          console.log('error getting info for engine', provider.name, error);
          return undefined;
        }
      }).filter(async info => info),
    );
    return infos;
```

Or do I misunderstand what you propose?

Comment (benoitf) [reply packages/main/src/plugin/container-registry.ts]: the trick is to use the cast after the filter call as you know it's filtering out undefined

edit: sorry it's not with as it's with `is` keyword

Comment (benoitf) [reply packages/main/src/plugin/container-registry.ts]: `filter((info): info is ContainerEngineInfo => !!info)`

**Review by benoitf — APPROVED** [review]

File: packages/main/src/plugin/container-registry.ts [hunk 3]
```diff
@@ -2334,12 +2337,35 @@ export class ContainerProviderRegistry {
...
+      providers = Array.from(this.internalProviders.values());
+    } else {
+      providers = [this.getMatchingContainerProvider(options?.provider)];
+    }
+    const infos = await Promise.all(
+      Array.from(providers).map(async provider => {
+        try {
+          return await this.info(provider.id);
+        } catch (error) {
+          console.log('error getting info for engine', provider.name, error);
```
Comment (axel7083) [inline packages/main/src/plugin/container-registry.ts]:
```suggestion
          console.error('error getting info for engine', provider.name, error);
```

**Review by axel7083 — APPROVED** [review]
LGTM!

**Comment by feloy** [comment]
I'm trying to call the method from an extension (one provided with Podman Desktop, kube-context), but I'm getting this error:

```
16:45:13 [main] Activating extension podman-desktop.kube-context failed error:TypeError: extensionApi__namespace.containerEngine.listInfos is not a function
```

Do I need to rebuild / copy the .d.ts file somewhere?

**Comment by benoitf** [comment]
@feloy `yarn build:preload:types`

**Comment by benoitf** [comment]
hmm no it should even work without that command

**Comment by feloy** [comment]
> hmm no it should even work without that command

It seems it fixed the problem. In any case, it works now. Thanks

**Pull request merged**